src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/debugger/DebuggerBase.java

```diff
@@ -232,6 +234,19 @@ protected final byte[] readBytes(long address, long numBytes)
     }
   }
 
+  /** If an address for a 64-bit value starts on the last 32-bit word of a
+      page, then we can't use the page cache to read it because it will cause
+      an ArrayIndexOutOfBoundsException when reading past the end of the page. */
+  private boolean canUsePageCacheFor64bitRead(long address) {
+    long pageMask = ~(pageSize - 1);
+    if ((address & pageMask) != ((address + 4) & pageMask)) {
```

Comment:
This looks a bit over-complicated.
Maybe something like
`    long pageMask = pageSize - 1;
    if ((address & pageMask) > (pageSize - 8)) {
      return false;
    }`

Reply:
I tinkered with this code quite a bit before I had something that both worked and I liked. I don't know that your approach is actually any simpler, and at the moment I still prefer mine, but that might just be because I'm more familiar with it. However, it does seem that there should be a more elegant way of doing this.

Reply:
Your code works only for the exact case - address of 64bit value points to the last 32bit of the page.

My idea is to use more generic approach - check that the whole value is in the page, i.e. this is special case of more common
```
boolean canUsePageCacheForRead(long address, int bits) {
   long pageMask = pageSize - 1;
   return ((address & pageMask) <= (pageSize - bits/8));
}

```
But I have no strong objections, feel free to keep your version

Reply:
> Your code works only for the exact case - address of 64bit value points to the last 32bit of the page.
That's the only case we need to worry about. Any others would have already failed the alignment check.